pkg/controllers/management/auth/globalroles/globalrole_handler_test.go

```diff
@@ -82,18 +138,367 @@ var (
+			wantError:  false,
+			condition: reducedCondition{
+				reason: ClusterRoleExists,
+				status: metav1.ConditionTrue,
+			},
+		},
+		{
+			name: "clusterRole is updated",
+			stateSetup: func(state grTestState) {
+				state.crListerMock.GetFunc = func(_, _ string) (*normanv1.ClusterRole, error) {
+					return readConfigCR.DeepCopy(), nil
+				}
+				state.crClientMock.UpdateFunc = func(cr *normanv1.ClusterRole) (*normanv1.ClusterRole, error) {
+					state.stateChanges.createdClusterRoles[cr.Name] = cr
+					return nil, nil
+				}
+			},
+			stateAssertions: func(gtsc grTestStateChanges) {
+				require.Len(gtsc.t, gtsc.createdClusterRoles, 1)
+				cr, ok := gtsc.createdClusterRoles["clusterRole"]
+				require.True(gtsc.t, ok)
+				require.Equal(gtsc.t, &readPodCR, cr)
+			},
+			globalRole: defaultGR.DeepCopy(),
+			wantError:  false,
+			condition: reducedCondition{
+				reason: ClusterRoleExists,
+				status: metav1.ConditionTrue,
+			},
+		},
+		{
+			name: "update clusterRole fails",
+			stateSetup: func(state grTestState) {
+				state.crListerMock.GetFunc = func(_, _ string) (*normanv1.ClusterRole, error) {
+					return readConfigCR.DeepCopy(), nil
+				}
+				state.crClientMock.UpdateFunc = func(cr *normanv1.ClusterRole) (*normanv1.ClusterRole, error) {
+					return nil, fmt.Errorf("error")
+				}
+			},
+			globalRole: defaultGR.DeepCopy(),
+			wantError:  true,
+			condition: reducedCondition{
+				reason: FailedToUpdateClusterRole,
+				status: metav1.ConditionFalse,
+			},
+		},
+		{
+			name: "create clusterRole fails",
+			stateSetup: func(state grTestState) {
+				state.crListerMock.GetFunc = func(_, _ string) (*normanv1.ClusterRole, error) {
+					return nil, nil
+				}
+				state.crClientMock.CreateFunc = func(cr *normanv1.ClusterRole) (*normanv1.ClusterRole, error) {
+					return nil, fmt.Errorf("error")
+				}
+			},
+			globalRole: defaultGR.DeepCopy(),
+			wantError:  true,
+			condition: reducedCondition{
+				reason: FailedToCreateClusterRole,
+				status: metav1.ConditionFalse,
+			},
+		},
+		{
+			name: "clusterRole is created",
+			stateSetup: func(state grTestState) {
+				state.crListerMock.GetFunc = func(_, _ string) (*normanv1.ClusterRole, error) {
+					return nil, nil
+				}
+				state.crClientMock.CreateFunc = func(cr *normanv1.ClusterRole) (*normanv1.ClusterRole, error) {
+					state.stateChanges.createdClusterRoles[cr.Name] = cr
+					return nil, nil
+				}
+			},
+			stateAssertions: func(gtsc grTestStateChanges) {
+				require.Len(gtsc.t, gtsc.createdClusterRoles, 1)
+				cr, ok := gtsc.createdClusterRoles["cattle-globalrole-"]
+				require.True(gtsc.t, ok)
+				require.Equal(gtsc.t, readPodCR.Rules, cr.Rules)
+			},
+			globalRole: defaultGR.DeepCopy(),
+			wantError:  false,
+			condition: reducedCondition{
+				reason: ClusterRoleExists,
+				status: metav1.ConditionTrue,
+			},
+			annotation: "cattle-globalrole-",
+		},
+	}
+
+	for _, test := range tests {
+		test := test
```

Comment:
That's a behaviour of Go, and how the `test` var is assigned. https://github.com/golang/go/issues/20733
I don't remember the internals, but it will be something fixed in the next version I guess.

- https://github.com/golang/go/discussions/56010
- https://github.com/golang/go/issues/60078
- https://go.dev/doc/faq#closures_and_goroutines

Not sure if all the links are related.

Reply:
Ah, the `test` in the  `for ...` line is scoped outside the loop, and `test` inside is then a loop-local (per iteration) shadow/copy. Thanks.

Reply:
Correct. I wasn't aware that it was going to be fixed in a future version of go. That would be nice! But for now it's an awkward necessity